rust-runtime/aws-smithy-types/src/lib.rs

```diff
@@ -80,49 +80,455 @@ pub enum Document {
 /// <https://docs.serde.rs/src/serde_json/number.rs.html#20-22>
 #[derive(Debug, Clone, Copy, PartialEq)]
 pub enum Number {
-    /// Unsigned 64-bit integer value
+    /// Unsigned 64-bit integer value.
     PosInt(u64),
-    /// Signed 64-bit integer value
+    /// Signed 64-bit integer value. The wrapped value is _always_ negative.
     NegInt(i64),
-    /// 64-bit floating-point value
+    /// 64-bit floating-point value.
     Float(f64),
 }
 
-macro_rules! to_num_fn {
-    ($name:ident, $typ:ident, $styp:expr) => {
+/* ANCHOR_END: document */
+
+impl Number {
+    /// Converts to an `f64` lossily.
+    /// Use `Number::try_from` to make the conversion only if it is not lossy.
+    pub fn to_f64_lossy(self) -> f64 {
+        match self {
+            Number::PosInt(v) => v as f64,
+            Number::NegInt(v) => v as f64,
+            Number::Float(v) => v as f64,
+        }
+    }
+
+    /// Converts to an `f32` lossily.
+    /// Use `Number::try_from` to make the conversion only if it is not lossy.
+    pub fn to_f32_lossy(self) -> f32 {
+        match self {
+            Number::PosInt(v) => v as f32,
+            Number::NegInt(v) => v as f32,
+            Number::Float(v) => v as f32,
+        }
+    }
+}
+
+/// The error type returned when conversion into an integer type or floating point type is lossy.
+#[derive(Debug)]
+pub enum TryFromNumberError {
+    /// Used when the conversion from an integer type into a smaller integer type would be lossy.
+    OutsideIntegerRange(std::num::TryFromIntError),
+    /// Used when the conversion from an `u64` into a floating point type would be lossy.
+    U64ToFloatLossyConversion(u64),
+    /// Used when the conversion from an `i64` into a floating point type would be lossy.
+    I64ToFloatLossyConversion(i64),
+    /// Used when attempting to convert an `f64` into an `f32`.
+    F64ToF32LossyConversion(f64),
+    /// Used when attempting to convert a decimal, infinite, or `NaN` floating point type into an
+    /// integer type.
+    UnexpectedFloat(f64),
+    /// Used when attempting to convert a negative [`Number`] into an unsigned integer type.
+    UnexpectedNegativeInteger(i64),
```

Comment:
I think these should get renamed to something like `FloatToIntegerLossyConversion` and `NegativeToUnsignedLossyConversion`, since "unexpected" really only applies to deserialization.

Reply:
https://github.com/awslabs/smithy-rs/pull/1274/commits/e17da696d2b01e44dd7031df00150bad6e87335b